physical/gcs.go

```diff
@@ -0,0 +1,182 @@
+package physical
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/mgutz/logxi/v1"
+
+	"cloud.google.com/go/storage"
+	"github.com/armon/go-metrics"
+	"golang.org/x/net/context"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+// GCSBackend is a physical backend that stores data
+// within an Google Cloud Storage bucket.
+type GCSBackend struct {
+	bucketName string
+	client     *storage.Client
+	logger     log.Logger
+}
+
+// newGCSBackend constructs a Google Cloud Storage backend using a pre-existing
+// bucket. Credentials can be provided to the backend, sourced
+// from environment variables or a service account file
+func newGCSBackend(conf map[string]string, logger log.Logger) (Backend, error) {
+
+	bucketName := os.Getenv("GOOGLE_STORAGE_BUCKET")
+
+	if bucketName == "" {
+		bucketName = conf["bucket"]
+		if bucketName == "" {
+			return nil, fmt.Errorf("'bucket' must be set")
+		}
+	}
+
+	// path to service account JSON file
+	credentialsFile := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
+	if credentialsFile == "" {
+		credentialsFile = conf["credentials_file"]
+		if credentialsFile == "" {
+			return nil, fmt.Errorf("no credentials specified for google storage")
+		}
+	}
+
+	client, err := storage.NewClient(
+		context.Background(),
+		option.WithServiceAccountFile(credentialsFile),
+	)
+
+	if err != nil {
+		return nil, fmt.Errorf("error establishing google storage client: '%v'", err)
+	}
+
+	// check client connectivity by getting bucket attributes
+	_, err = client.Bucket(bucketName).Attrs(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("unable to access bucket '%s': '%v'", bucketName, err)
+	}
+
+	s := GCSBackend{
+		client:     client,
+		bucketName: bucketName,
+		logger:     logger,
+	}
+
+	return &s, nil
+}
+
+// Put is used to insert or update an entry
+func (s *GCSBackend) Put(entry *Entry) error {
+	defer metrics.MeasureSince([]string{"google storage", "put"}, time.Now())
+
+	bucket := s.client.Bucket(s.bucketName)
+	writer := bucket.Object(entry.Key).NewWriter(context.Background())
+
+	_, err := writer.Write(entry.Value)
+	if cerr := writer.Close(); cerr != nil && err == nil {
```

Comment:
I'm not in favor of returning an error on close. The reason is that doing so will prevent various other parts of Vault (such as the read cache) from acknowledging that the operation was successful, even though the write itself was (unless the write may not be successful with GCS until Close is called). This can then cause the cache and underlying data store to think that the contained values are different.

I would instead propose simply logging errors to the logger if you see them on close.